# Web Scraping Lab

You will find in this notebook some web scraping exercises to practice your scraping skills using `requests` and `Beautiful Soup`.

**Tips:**

- Check the [response status code](https://http.cat/) for each request to ensure you have obtained the intended content.
- Look at the HTML code in each request to understand the kind of info you are getting and its format.
- Check for patterns in the response text to extract the data/info requested in each question.
- Visit each url and take a look at its source through Chrome DevTools. You'll need to identify the html tags, special class names etc. used for the html content you are expected to extract.
- Check out the css selectors.

### Useful Resources
- [Requests library](http://docs.python-requests.org/en/master/#the-user-guide) documentation 
- [Beautiful Soup Doc](https://www.crummy.com/software/BeautifulSoup/bs4/doc/)
- [List of HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes)
- [HTML basics](http://www.simplehtmlguide.com/cheatsheet.php)
- [CSS basics](https://www.cssbasics.com/#page_start)

#### First of all, gathering our tools.

In [1]:
import requests
from bs4 import BeautifulSoup
import pandas as pd
import json

⚠️ **Again, please remember to limit your output before submission so that your code doesn't get lost in the output.**

#### Challenge 1 - Download, parse (using BeautifulSoup), and print the content from the Trending Developers page from GitHub:

In [2]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/developers'

In [3]:
response = requests.get(url)

In [4]:
response.status_code

200

In [5]:
soup = BeautifulSoup(response.content)

In [6]:
print(soup.prettify())

<!DOCTYPE html>
<html lang="en">
 <head>
  <meta charset="utf-8"/>
  <link href="https://github.githubassets.com" rel="dns-prefetch"/>
  <link href="https://avatars.githubusercontent.com" rel="dns-prefetch"/>
  <link href="https://github-cloud.s3.amazonaws.com" rel="dns-prefetch"/>
  <link href="https://user-images.githubusercontent.com/" rel="dns-prefetch"/>
  <link crossorigin="anonymous" href="https://github.githubassets.com/assets/frameworks-6ef98f9c87d25ee43b6e8b15f18b3e5c.css" integrity="sha512-bvmPnIfSXuQ7bosV8Ys+XNjZ7wqcfxwStr1JGjt4ibr/TK0ycsQZAYi4/4aziAgybJgv2/5+bw4lJF7CRSRs/g==" media="all" rel="stylesheet"/>
  <link crossorigin="anonymous" href="https://github.githubassets.com/assets/site-6547fa2bebd0a8c63e461573691982cb.css" integrity="sha512-ZUf6K+vQqMY+RhVzaRmCy2ePbSZad4TkaGRbd6v5gFt6f9Q/nqjDkBDjQgXNmZw7J9mcYxlsE4fhRw7CTluRow==" media="all" rel="stylesheet"/>
  <link crossorigin="anonymous" href="https://github.githubassets.com/assets/behaviors-9a6d928a0cc4b9d03dc52f27ca2

In [7]:
len(soup)

3

#### Display the names of the trending developers retrieved in the previous step.

Your output should be a Python list of developer names. Each name should not contain any html tag.

**Instructions:**

1. Find out the html tag and class names used for the developer names. You can achieve this using Chrome DevTools.

1. Use BeautifulSoup to extract all the html elements that contain the developer names.

1. Use string manipulation techniques to replace whitespaces and linebreaks (i.e. `\n`) in the *text* of each html element. Use a list to store the clean names.

1. Print the list of names.

Your output should look like below (with different names):

```
['trimstray (@trimstray)',
 'joewalnes (JoeWalnes)',
 'charlax (Charles-AxelDein)',
 'ForrestKnight (ForrestKnight)',
 'revery-ui (revery-ui)',
 'alibaba (Alibaba)',
 'Microsoft (Microsoft)',
 'github (GitHub)',
 'facebook (Facebook)',
 'boazsegev (Bo)',
 'google (Google)',
 'cloudfetch',
 'sindresorhus (SindreSorhus)',
 'tensorflow',
 'apache (TheApacheSoftwareFoundation)',
 'DevonCrawford (DevonCrawford)',
 'ARMmbed (ArmMbed)',
 'vuejs (vuejs)',
 'fastai (fast.ai)',
 'QiShaoXuan (Qi)',
 'joelparkerhenderson (JoelParkerHenderson)',
 'torvalds (LinusTorvalds)',
 'CyC2018',
 'komeiji-satori (神楽坂覚々)',
 'script-8']
 ```

To find the names:

In [8]:
soup.select("h1.h3.lh-condensed a ")[0].text


'\n            Michele Locati\n'

To find the nick

In [9]:
soup.select("p a")[0].text

'\n              mlocati\n'

In [10]:
def extract_name(soup):
    return soup.select("h1.h3.lh-condensed a")[0].text
def extract_nick(soup):
    return soup.select("p a")[0].text

In [11]:
nombres=soup.select("h1.h3.lh-condensed a")

In [12]:
len(nombres)

25

In [13]:
nicks=soup.select("p a")

In [14]:
len(nicks)

25

In [15]:
box=soup.select("article.Box-row.d-flex")

In [16]:
len(box)

25

In [17]:
name_nicks = [(extract_name(n).strip(), extract_nick(n).strip()) for n in box]

In [18]:
name_nicks[:10]

[('Michele Locati', 'mlocati'),
 ('Han Xiao', 'hanxiao'),
 ('Alon Zakai', 'kripken'),
 ('Ha Thach', 'hathach'),
 ('Joel Arvidsson', 'oblador'),
 ('Major Hayden', 'major'),
 ('Ahmet Alp Balkan', 'ahmetb'),
 ('Don Jayamanne', 'DonJayamanne'),
 ('Alexey Golub', 'Tyrrrz'),
 ('Diego Muracciole', 'diegomura')]

####  Challenge 2 - Display the trending Python repositories in GitHub

The steps to solve this problem is similar to the previous one except that you need to find out the repository names instead of developer names.

In [19]:
# This is the url you will scrape in this exercise
url = 'https://github.com/trending/python?since=daily'

In [20]:
response = requests.get(url)
soup = BeautifulSoup(response.content)
name_repos=soup.select("h1.h3.lh-condensed a")

In [21]:
name_repos[0].text.strip().replace("\n","").replace(" ","")

'jina-ai/jina'

In [22]:
repos=[repo.text.strip().replace("\n","").replace(" ","") for repo in name_repos]

In [23]:
reposplit = [element.split("/") for element in repos]

In [64]:
repo_final = [element[1] for element in reposplit]

In [65]:
repo_final[:10]

['jina',
 'bypass_utility',
 'awx',
 '30-Days-Of-Python',
 'PaddleOCR',
 'wsbtickerbot',
 'detectron2',
 'public-apis',
 'transformers',
 'CheatSheetSeries']

#### Challenge 3 - Display all the image links from Walt Disney wikipedia page

In [25]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/Walt_Disney'

In [26]:
response = requests.get(url)
soup = BeautifulSoup(response.content)
img_tags = soup.find_all("img")

In [27]:
len(img_tags)

34

In [28]:
img_tags[0].get("src")

'//upload.wikimedia.org/wikipedia/en/thumb/e/e7/Cscr-featured.svg/20px-Cscr-featured.svg.png'

In [29]:
links_images=[fig.get("src") for fig in img_tags]

In [30]:
links_images[:10]

['//upload.wikimedia.org/wikipedia/en/thumb/e/e7/Cscr-featured.svg/20px-Cscr-featured.svg.png',
 '//upload.wikimedia.org/wikipedia/en/thumb/8/8c/Extended-protection-shackle.svg/20px-Extended-protection-shackle.svg.png',
 '//upload.wikimedia.org/wikipedia/commons/thumb/d/df/Walt_Disney_1946.JPG/220px-Walt_Disney_1946.JPG',
 '//upload.wikimedia.org/wikipedia/commons/thumb/8/87/Walt_Disney_1942_signature.svg/150px-Walt_Disney_1942_signature.svg.png',
 '//upload.wikimedia.org/wikipedia/commons/thumb/c/c4/Walt_Disney_envelope_ca._1921.jpg/220px-Walt_Disney_envelope_ca._1921.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/4/4d/Newman_Laugh-O-Gram_%281921%29.webm/220px-seek%3D2-Newman_Laugh-O-Gram_%281921%29.webm.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/0/0d/Trolley_Troubles_poster.jpg/170px-Trolley_Troubles_poster.jpg',
 '//upload.wikimedia.org/wikipedia/commons/thumb/7/71/Walt_Disney_and_his_cartoon_creation_%22Mickey_Mouse%22_-_National_Board_of_Review_Magazine.jpg/170

#### Challenge 4 - Retrieve all links to pages on Wikipedia that refer to some kind of Python.

In [31]:
# This is the url you will scrape in this exercise
url ='https://en.wikipedia.org/wiki/Python' 
response = requests.get(url)
soup = BeautifulSoup(response.content)
links = soup.select("div.mw-parser-output ul li a")

In [32]:
len(links)

33

In [33]:
list_links=[n.get("href") for n in links]

In [34]:
final_list=[]
for i in list_links:
    if "#" not in i:
        final_list.append(i)
final_list

['/wiki/Pythons',
 '/wiki/Python_(genus)',
 '/wiki/Python_(programming_language)',
 '/wiki/CMU_Common_Lisp',
 '/wiki/Python_of_Aenus',
 '/wiki/Python_(painter)',
 '/wiki/Python_of_Byzantium',
 '/wiki/Python_of_Catana',
 '/wiki/Python_Anghelo',
 '/wiki/Python_(Efteling)',
 '/wiki/Python_(Busch_Gardens_Tampa_Bay)',
 '/wiki/Python_(Coney_Island,_Cincinnati,_Ohio)',
 '/wiki/Python_(automobile_maker)',
 '/wiki/Python_(Ford_prototype)',
 '/wiki/Python_(missile)',
 '/wiki/Python_(nuclear_primary)',
 '/wiki/Colt_Python',
 '/wiki/PYTHON',
 '/wiki/Python_(film)',
 '/wiki/Python_(mythology)',
 '/wiki/Monty_Python',
 '/wiki/Python_(Monty)_Pictures',
 '/wiki/Cython',
 '/wiki/Pyton',
 '/wiki/Pithon']

#### Challenge 5 - Number of Titles that have changed in the United States Code since its last release point 

In [35]:
# This is the url you will scrape in this exercise
url = 'http://uscode.house.gov/download/download.shtml'

In [36]:
response = requests.get(url)
soup = BeautifulSoup(response.content)
tittles_changed = soup.select("div.usctitlechanged")
print(len(tittles_changed), "titles changed after last revision.")

12 titles changed after last revision.


#### Challenge 6 - A Python list with the top ten FBI's Most Wanted names 

In [37]:
# This is the url you will scrape in this exercise
url = 'https://www.fbi.gov/wanted/topten'

In [38]:
response = requests.get(url)
response.status_code


503

Probably the website is protected...¿?

#### Challenge 7 - List all language names and number of related articles in the order they appear in wikipedia.org

In [39]:
# This is the url you will scrape in this exercise
url = 'https://www.wikipedia.org/'

In [40]:
response = requests.get(url)
soup = BeautifulSoup(response.content)
lang_packs = soup.select('div.central-featured div')


In [41]:
len(lang_packs)

10

In [42]:
soup.select("a.link-box strong")[9].text

'Português'

In [43]:
soup.select("a.link-box small bdi")[9].text

'1\xa0053\xa0000+'

In [44]:
def language_get(soup):
    return soup.select("a.link-box strong")[0].text
def related_articles(soup):
    string = soup.select("a.link-box small bdi")[0].text
    number=string.replace("\xa0",",").replace("+","")
    return number


In [45]:
lang_list = [[language_get(lang),related_articles(lang)] for lang in lang_packs]

In [46]:
lang_list

[['English', '6,235,000'],
 ['日本語', '1,250,000'],
 ['Deutsch', '2,528,000'],
 ['Español', '1,655,000'],
 ['Русский', '1,693,000'],
 ['Français', '2,292,000'],
 ['Italiano', '1,669,000'],
 ['中文', '1,172,000'],
 ['Polski', '1,452,000'],
 ['Português', '1,053,000']]

#### Challenge 8 - A list with the different kind of datasets available in data.gov.uk 

In [47]:
# This is the url you will scrape in this exercise
url = 'https://data.gov.uk/'

In [48]:
response = requests.get(url)
soup = BeautifulSoup(response.content)
gov_uk = soup.select('li h3 a')

In [49]:
list_data_uk = [data.text for data in gov_uk]

In [50]:
list_data_uk

['Business and economy',
 'Crime and justice',
 'Defence',
 'Education',
 'Environment',
 'Government',
 'Government spending',
 'Health',
 'Mapping',
 'Society',
 'Towns and cities',
 'Transport']

#### Challenge 9 - Top 10 languages by number of native speakers stored in a Pandas Dataframe

Este no me sale, y no entiendo por qué, las búsquedas que estoy realizando en el select me funcionan en Google Chrome, sin embargo aquí no me cogen nada...

In [51]:
# This is the url you will scrape in this exercise
url = 'https://en.wikipedia.org/wiki/List_of_languages_by_number_of_native_speakers'

In [52]:
response = requests.get(url)
soup = BeautifulSoup(response.content)
languages = soup.select('table.wikitable.sortable.jquery-tablesorter:nth-of-type(4) tbody')

In [53]:
languages

[]

In [54]:
def language_get(soup):
    return soup.select("td a")[0].text
def speakers_get(soup):
    string = soup.select("td:first-child+td+td")[0].text

In [55]:
list_languages = [(language_get(item),speakers_get(item))for item in languages]

In [56]:
list_languages[:10]

[]

### Stepping up the game

####  Challenge 10 - The 20 latest earthquakes info (date, time, latitude, longitude and region name) by the EMSC as a pandas dataframe

In [57]:
# This is the url you will scrape in this exercise
url = 'https://www.emsc-csem.org/Earthquake/'

#### Challenge 11 - IMDB's Top 250 data (movie name, Initial release, director name and stars) as a pandas dataframe

In [58]:
# This is the url you will scrape in this exercise 
url = 'https://www.imdb.com/chart/top'

#### Challenge 12 - Movie name, year and a brief summary of the top 10 random movies (IMDB) as a pandas dataframe.

In [59]:
#This is the url you will scrape in this exercise
url = 'http://www.imdb.com/chart/top'

#### Challenge 13 - Find the live weather report (temperature, wind speed, description and weather) of a given city.

In [60]:
#https://openweathermap.org/current

In [61]:
def weather(city):
    pass

#### Challenge 14 - Book name,price and stock availability as a pandas dataframe.

In [62]:
# This is the url you will scrape in this exercise. 
# It is a fictional bookstore created to be scraped. 
url = 'http://books.toscrape.com/'

**Did you limit your output? Thank you! 🙂**